spring-amqp/src/main/java/org/springframework/amqp/support/converter/DefaultJackson2JavaTypeMapper.java

```diff
@@ -29,49 +31,100 @@
  * @author Sam Nelson
  * @author Andreas Asplund
  * @author Artem Bilan
+ * @author Gary Russell
  */
-public class DefaultJackson2JavaTypeMapper extends AbstractJavaTypeMapper implements Jackson2JavaTypeMapper, ClassMapper {
+public class DefaultJackson2JavaTypeMapper extends AbstractJavaTypeMapper
+		implements Jackson2JavaTypeMapper, ClassMapper {
+
+	private static final JavaType STRING_TYPE = TypeFactory.defaultInstance().constructType(String.class);
+
+	private static final JavaType OBJECT_TYPE = TypeFactory.defaultInstance().constructType(Object.class);
+
+	private volatile TypePrecedence typePrecedence = TypePrecedence.INFERRED;
+
+	/**
+	 * @return the precedence.
+	 * @see #setTypePrecedence(Jackson2JavaTypeMapper.TypePrecedence)
+	 * @since 1.6.
+	 */
+	@Override
+	public TypePrecedence getTypePrecedence() {
+		return this.typePrecedence;
+	}
+
+	/**
+	 * Set the precedence for evaluating type information in message properties.
+	 * When using {@code @RabbitListener} at the method level, the framework attempts
+	 * to determine the target type for payload conversion from the method signature.
+	 * If so, this type is provided in the
+	 * {@link MessageProperties#getInferredArgumentType() inferredArgumentType}
+	 * message property.
+	 * <p>
+	 * By default, if the type is concrete (not abstract, not an interface), this will
+	 * be used ahead of type information provided in the {@code __TypeId__} and
+	 * associated headers provided by the sender.
+	 * <p>
+	 * If you wish to force the use of the  {@code __TypeId__} and associated headers
+	 * (such as when the actual type is a subclass of the method argument type),
+	 * set the precedence to {@link Jackson2JavaTypeMapper.TypePrecedence#TYPE_ID}.
+	 *
+	 * @param typePrecedence the precedence.
+	 * @since 1.6
+	 */
+	public void setTypePrecedence(TypePrecedence typePrecedence) {
+		this.typePrecedence = typePrecedence;
```

Comment:
`Assert.notNull(typePrecedence, "'typePrecedence' must not be null.");`
Otherwise NPE for the `equals` operation in the `toJavaType()`.